File: backend/app/models/budget.py
from sqlalchemy import Column, Integer, String, Float, ForeignKey, DateTime, Enum
from sqlalchemy.orm import relationship
from datetime import datetime
import enum
from app.models.base import Base
from app.models.user import User

class BudgetPeriod(str, enum.Enum):
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    YEARLY = "yearly"

class BudgetCategory(str, enum.Enum):
    HOUSING = "housing"
    TRANSPORTATION = "transportation"
    FOOD = "food"
    UTILITIES = "utilities"
    HEALTHCARE = "healthcare"
    ENTERTAINMENT = "entertainment"
    SHOPPING = "shopping"
    SAVINGS = "savings"
    DEBT = "debt"
    OTHER = "other"

class Budget(Base):
    __tablename__ = "budgets"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    name = Column(String, nullable=False)
    category = Column(Enum(BudgetCategory), nullable=False)
    amount = Column(Float, nullable=False)
    period = Column(Enum(BudgetPeriod), nullable=False)
    start_date = Column(DateTime, nullable=False)
    end_date = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    user = relationship("User", back_populates="budgets")
    expenses = relationship("BudgetExpense", back_populates="budget")
    alerts = relationship("BudgetAlert", back_populates="budget")

class BudgetExpense(Base):
    __tablename__ = "budget_expenses"

    id = Column(Integer, primary_key=True, index=True)
    budget_id = Column(Integer, ForeignKey("budgets.id"), nullable=False)
    transaction_id = Column(Integer, ForeignKey("transactions.id"), nullable=True)
    amount = Column(Float, nullable=False)
    description = Column(String, nullable=True)
    date = Column(DateTime, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)

    # Relationships
    budget = relationship("Budget", back_populates="expenses")
    transaction = relationship("Transaction")

class AlertType(str, enum.Enum):
    THRESHOLD = "threshold"
    OVERSPENT = "overspent"
    APPROACHING_LIMIT = "approaching_limit"
    RECURRING_DUE = "recurring_due"

class AlertStatus(str, enum.Enum):
    ACTIVE = "active"
    DISMISSED = "dismissed"
    RESOLVED = "resolved"

class BudgetAlert(Base):
    __tablename__ = "budget_alerts"

    id = Column(Integer, primary_key=True, index=True)
    budget_id = Column(Integer, ForeignKey("budgets.id"), nullable=False)
    type = Column(Enum(AlertType), nullable=False)
    status = Column(Enum(AlertStatus), default=AlertStatus.ACTIVE)
    message = Column(String, nullable=False)
    threshold_percentage = Column(Float, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    budget = relationship("Budget", back_populates="alerts")

# Update User model to include budget relationship
User.budgets = relationship("Budget", back_populates="user")

File: frontend/src/types/budget.ts
export enum BudgetCategory {
  HOUSING = 'housing',
  TRANSPORTATION = 'transportation',
  FOOD = 'food',
  UTILITIES = 'utilities',
  HEALTHCARE = 'healthcare',
  ENTERTAINMENT = 'entertainment',
  SHOPPING = 'shopping',
  SAVINGS = 'savings',
  DEBT = 'debt',
  OTHER = 'other',
}

export enum BudgetPeriod {
  DAILY = 'daily',
  WEEKLY = 'weekly',
  MONTHLY = 'monthly',
  YEARLY = 'yearly',
}

export enum AlertType {
  THRESHOLD = 'threshold',
  OVERSPENT = 'overspent',
  APPROACHING_LIMIT = 'approaching_limit',
  RECURRING_DUE = 'recurring_due',
}

export enum AlertStatus {
  ACTIVE = 'active',
  DISMISSED = 'dismissed',
  RESOLVED = 'resolved',
}

export interface Budget {
  id: number;
  userId: number;
  name: string;
  category: BudgetCategory;
  amount: number;
  period: BudgetPeriod;
  startDate: string;
  endDate?: string;
  spent?: number;
  remaining?: number;
  createdAt: string;
  updatedAt: string;
}

export interface BudgetCreate {
  name: string;
  category: BudgetCategory;
  amount: number;
  period: BudgetPeriod;
  startDate: string;
  endDate?: string;
}

export interface BudgetUpdate {
  name?: string;
  category?: BudgetCategory;
  amount?: number;
  period?: BudgetPeriod;
  endDate?: string;
}

export interface BudgetExpense {
  id: number;
  budgetId: number;
  transactionId?: number;
  amount: number;
  description?: string;
  date: string;
  createdAt: string;
}

export interface BudgetExpenseCreate {
  transactionId?: number;
  amount: number;
  description?: string;
  date?: string;
}

export interface BudgetAlert {
  id: number;
  budgetId: number;
  type: AlertType;
  status: AlertStatus;
  message: string;
  thresholdPercentage?: number;
  createdAt: string;
  updatedAt: string;
}

export interface CategorySpending {
  totalSpent: number;
  budgetAmount: number;
}
